enterprise/server/execution_search_service/execution_search_service.go

```diff
@@ -0,0 +1,247 @@
+	rows, err := s.oh.RawWithOptions(ctx, clickhouse.Opts().WithQueryName("search_executions"), query, queryArgs...).Rows()
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	executions := make([]*schema.Execution, 0)
+	for rows.Next() {
+		var te schema.Execution
+		if err := s.oh.DB(ctx).ScanRows(rows, &te); err != nil {
+			return nil, err
+		}
+		executions = append(executions, &te)
+	}
+	return executions, nil
+}
+
+type ExecutionWithInvocationId struct {
+	execution    *expb.Execution
+	invocationID string
+}
+
+func (s *ExecutionSearchService) fetchExecutionData(ctx context.Context, groupId string, execIds []string) (map[string]*ExecutionWithInvocationId, error) {
+	qString := "SELECT * FROM Executions WHERE (execution_id IN ?) AND (perms & ? != 0) AND (group_id = ?)"
+	qArgs := []interface{}{execIds, perms.GROUP_READ, groupId}
+
+	rows, err := s.h.RawWithOptions(ctx, db.Opts().WithQueryName("fetch_executions"), qString, qArgs...).Rows()
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	executions := make(map[string]*ExecutionWithInvocationId, 0)
+	for rows.Next() {
+		var r tables.Execution
+		if err := s.h.DB(ctx).ScanRows(rows, &r); err != nil {
+			return nil, err
+		}
+		exec, err := execution.TableExecToClientProto(&r)
+		if err != nil {
+			return nil, err
+		}
+		executions[r.ExecutionID] = &ExecutionWithInvocationId{execution: exec, invocationID: r.InvocationID}
+	}
+	return executions, nil
+}
+
+func clickhouseExecutionToProto(in *schema.Execution, ex *ExecutionWithInvocationId) (*expb.ExecutionWithInvocationMetadata, error) {
+	if in == nil || ex == nil {
+		return nil, status.InternalErrorf("Execution not found or not accessible.")
+	}
+	return &expb.ExecutionWithInvocationMetadata{
+		Execution: ex.execution,
+		InvocationMetadata: &expb.InvocationMetadata{
+			Id:               ex.invocationID,
+			User:             in.User,
+			Host:             in.Host,
+			Pattern:          in.Pattern,
+			Role:             in.Role,
+			BranchName:       in.BranchName,
+			CommitSha:        in.CommitSHA,
+			RepoUrl:          in.RepoURL,
+			Command:          in.Command,
+			Success:          in.Success,
+			InvocationStatus: ispb.InvocationStatus(in.InvocationStatus),
+		},
+	}, nil
+}
+
+func (s *ExecutionSearchService) SearchExecutions(ctx context.Context, req *expb.SearchExecutionRequest) (*expb.SearchExecutionResponse, error) {
+	if s.oh == nil {
+		return nil, status.UnavailableError("An OLAP DB is required to search executions.")
+	}
+	u, err := perms.AuthenticatedUser(ctx, s.env)
+	if err != nil {
+		return nil, err
+	}
+	if u.GetGroupID() == "" {
+		return nil, status.InvalidArgumentError("Failed to find user's group when searching executions.")
+	}
+	if err := perms.AuthorizeGroupAccessForStats(ctx, s.env, u.GetGroupID()); err != nil {
+		return nil, err
+	}
+
+	q := query_builder.NewQuery(`SELECT * FROM Executions`)
```

Comment:
We have code in perms.go (AddPermissionsCheckToQuery) to automatically craft a permission filter clause. You should add that on top of whatever other filtering you are doing.

Reply:
There aren't any perms in the clickhouse table for executions, so what I've done here is added a filter to restrict to the authz'd group and then am also checking perms as part of the lookup in mysql--we don't return any results if we don't find even one of the executions in the permissions-guarded mysql query.  I started using AddPermissionsCheckToQuery there instead of what i'm using now (and i'm still including the group filter in both cases).